### PR TITLE
PR feedback on syncing.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/72990.

### CODE REVIEW

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs [hunk 1]
```diff
@@ -152,6 +149,18 @@ async Task<ImmutableArray<DocumentInfo>> CreateDocumentInfosAsync(DocumentChecks
         // TODO: do we need version?
         return new DocumentInfo(attributes, textLoader, documentServiceProvider: null);
     }
+
+    public AssetHelper<T> GetAssetHelper<T>()
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/AbstractAssetProvider.cs:153]: this pattern helped make the callsites work better with inference, by splitting the type args into the parts that have to be specified (the `T`) and the part that can be inferred (the `TArgs`).

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs [hunk 2]
```diff
@@ -293,47 +294,23 @@ public int Compare(DocumentId? x, DocumentId? y)
...
+            textChecksums[index] = stateChecksums.Text;
+            index++;
         }
 
         return new(
-            new ChecksumCollection(attributeChecksums.ToImmutableAndClear()),
-            new ChecksumCollection(textChecksums.ToImmutableAndClear()),
+            new ChecksumCollection(ImmutableCollectionsMarshal.AsImmutableArray(attributeChecksums)),
+            new ChecksumCollection(ImmutableCollectionsMarshal.AsImmutableArray(textChecksums)),
             SelectAsArray(static s => s.Id));
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs:314]: >SelectAsArray [](http://example.com/codeflow?start=12&length=13)

Were the SelectAsArray methods worth changing too?

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs [hunk 3]
```diff
@@ -152,6 +149,18 @@ async Task<ImmutableArray<DocumentInfo>> CreateDocumentInfosAsync(DocumentChecks
         // TODO: do we need version?
```
Comment (ToddGrun) [inline src/Workspaces/Remote/Core/AbstractAssetProvider.cs:149]: >// TODO: do we need version? [](http://example.com/codeflow?start=8&length=28)

TODO still in flux?

Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/Core/AbstractAssetProvider.cs:149]: want to talk to you in person about this.

File: src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs (hunk 2 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs]: Yup.  Fixed up.

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs [hunk 4]
```diff
@@ -188,14 +197,14 @@ internal static class AbstractAssetProviderExtensions
         using var _1 = PooledHashSet<Checksum>.GetInstance(out var checksumSet);
         checksumSet.AddAll(checksums.Children);
 
-        using var _2 = ArrayBuilder<T>.GetInstance(checksumSet.Count, out var builder);
+        var builder = ImmutableArray.CreateBuilder<T>(checksumSet.Count);
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/Workspaces/Remote/Core/AbstractAssetProvider.cs:200]: > This is probably better

This should be better.  An IAB starts out with an array of exactly the size you recommend.  And we then move that array directly to the IA result.  So the only thing that happens here is that builder alloc itself (just a tiny object) is garbage.

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs (hunk 3 above)
Comment (ToddGrun) [reply src/Workspaces/Remote/Core/AbstractAssetProvider.cs:149]: Might be a week or two before I'm back in the office due to that machine going down. I don't have any strong thoughts here, I just saw a todo and wondered if it was still valid.

File: src/Workspaces/Remote/Core/AbstractAssetProvider.cs [hunk 5]
```diff
@@ -188,14 +197,14 @@ internal static class AbstractAssetProviderExtensions
...
 
-        await assetProvider.GetAssetsAsync<T, ArrayBuilder<T>>(
+        await assetProvider.GetAssetHelper<T>().GetAssetsAsync(
             assetPath, checksumSet,
             static (checksum, asset, builder) => builder.Add(asset),
             builder,
             cancellationToken).ConfigureAwait(false);
 
-        return builder.ToImmutableAndClear();
+        return builder.MoveToImmutable();
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/Core/AbstractAssetProvider.cs:208]: @ToddGrun note the movetoimmutable here.  the builder itself produces no internal garbage.  it just holds onto the array and count.  MoveToImmtuable is just an ownership transform of its array to the IA result.  So the only thing left is the builder itself, which should be trivially collected.  

Note: one pattern we could consider here is pooling these builders *when* we've successfully moved their internal arrays out.

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs [hunk 6]
```diff
@@ -281,6 +280,7 @@ static SourceGeneratorExecutionVersionMap FilterToProjectCone(SourceGeneratorExe
                 // efficiently in bulk and in parallel.
                 await _assetProvider.SynchronizeProjectAssetsAsync(projectStateChecksumsToAdd, cancellationToken).ConfigureAwait(false);
 
+                using var _3 = ArrayBuilder<ProjectInfo>.GetInstance(out var projectInfos);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.SolutionCreator.cs:283]:
```suggestion
                using var _3 = ArrayBuilder<ProjectInfo>.GetInstance(projectStateChecksumsToAdd.Count, out var projectInfos);
```

**Review by ToddGrun — APPROVED** [review]
:shipit:

**Comment by CyrusNajmabadi** [comment]
@jasonmalinowski 	For review when you get back.

**Pull request merged**